### PR TITLE
sink(ticdc): remove useless buffer sink and sink manager

### PR DESCRIPTION
**Pull request by hi-rustin**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tiflow/issues/5283

### What is changed and how it works?

- remove useless buffer sink
- remove sink manager
- make dispatcher thread-safe
  - We will call EmitRowChangeToSink concurrently after we delete the bufferSink, so we need it to be thread-safe, otherwise there will be incorrect partition selection. This is the reason why we have been failing in kafka integration testing in the past.

### Check List 

Tests 

 - Unit test
 - Integration test
 - [Manual testing
](https://tcms.pingcap.net/dashboard/executions/plan/754783)

Code changes

None

Side effects

None

Related changes

None

### Release note 

```release-note
None
```

### CODE REVIEW

File: cdc/sink/sink_manager.go [hunk 1]
```diff
@@ -30,7 +30,7 @@ import (
 // and backendSink.
 // Manager is thread-safe.
 type Manager struct {
-	bufSink      *bufferSink
+	sink         Sink
```
Comment (CharlesCheung96) [inline cdc/sink/sink_manager.go:33]: Perhaps we could remove sink_manager and move the relevant fields to processor.

Comment (hi-rustin) [reply cdc/sink/sink_manager.go:33]: I guess if we merge it will increase the complexity? Is the processor already complex enough on its own?

**Comment by lonng** [comment]
Nice work！

**Comment by hi-rustin** [comment]
@amyangfei 
Could you please take a look? Thanks! 💚 💙 💜 💛 ❤️

File: cdc/sink/mq/dispatcher/partition/index_value.go [hunk 2]
```diff
@@ -14,13 +14,16 @@
...
+	"sync"
+
 	"github.com/pingcap/tiflow/cdc/model"
 	"github.com/pingcap/tiflow/pkg/hash"
 )
 
 // IndexValueDispatcher is a partition dispatcher which dispatches events based on the index value.
 type IndexValueDispatcher struct {
 	hasher *hash.PositionInertia
+	lock   sync.Mutex
```
Comment (liuzix) [inline cdc/sink/mq/dispatcher/partition/index_value.go:26]: Could you do a benchmark here proving that lock contention will not be a bottleneck?

File: cdc/sink/table_sink.go [hunk 3]
```diff
@@ -116,15 +137,15 @@ func (t *tableSink) EmitCheckpointTs(_ context.Context, _ uint64, _ []model.Tabl
...
 // Close once the method is called, no more events can be written to this table sink
 func (t *tableSink) Close(ctx context.Context) error {
-	return t.manager.destroyTableSink(ctx, t.tableID)
+	return t.Barrier(ctx, t.tableID)
 }
 
 // Barrier is not used in table sink
 func (t *tableSink) Barrier(ctx context.Context, tableID model.TableID) error {
-	return nil
+	return t.backendSink.Barrier(ctx, tableID)
```
Comment (amyangfei) [inline cdc/sink/table_sink.go:150]: Will the backend sink guarantee that no more data whose `commit-ts` is larger than the max resolved ts(which is decided by now) will be written to downstream after `Barrier` is called.

File: cdc/sink/mq/dispatcher/partition/index_value.go [hunk 4]
```diff
@@ -33,6 +36,8 @@ func NewIndexValueDispatcher() *IndexValueDispatcher {
 // DispatchRowChangedEvent returns the target partition to which
 // a row changed event should be dispatched.
 func (r *IndexValueDispatcher) DispatchRowChangedEvent(row *model.RowChangedEvent, partitionNum int32) int32 {
+	r.lock.Lock()
+	defer r.lock.Unlock()
 	r.hasher.Reset()
 	r.hasher.Write([]byte(row.Table.Schema), []byte(row.Table.Table))
```
Comment (amyangfei) [inline cdc/sink/mq/dispatcher/partition/index_value.go:42]: Does the lock protects access of `r.hasher` only?

File: cdc/sink/mq/dispatcher/partition/index_value.go (hunk 2 above)
Comment (hi-rustin) [reply cdc/sink/mq/dispatcher/partition/index_value.go:26]: Sure. I'll test it.

File: cdc/sink/table_sink.go (hunk 3 above)
Comment (hi-rustin) [reply cdc/sink/table_sink.go:150]: We paused the data to continue sending before we called barrier, so there should be no new data after we're done.
@sdojjy What do you think?

File: cdc/sink/mq/dispatcher/partition/index_value.go (hunk 4 above)
Comment (hi-rustin) [reply cdc/sink/mq/dispatcher/partition/index_value.go:42]: Yes, it protects the atomicity of these operations.

File: cdc/sink/mq/dispatcher/partition/index_value.go (hunk 2 above)
Comment (hi-rustin) [reply cdc/sink/mq/dispatcher/partition/index_value.go:26]: I think the impact in this area is limited, in this PR I am moving the locks from the original BufferSink to the dispatcher level. Maybe a real cluster test is needed to determine the impact on it.

**Comment by hi-rustin** [comment]
Performance tests:

```topology.yaml
global:
  user: "root"
  ssh_port: 22
  deploy_dir: "/tidb-deploy"
  arch: "amd64"
  resource_control:
     memory_limit: "12G"
     cpu_quota: "200%"

monitored:
  node_exporter_port: 9100
  blackbox_exporter_port: 9115

pd_servers:
  - host: vsc.liudongpo.tispace
  - host: test1.liudongpo.tispace
  - host: test2.liudongpo.tispace

tidb_servers:
  - host: vsc.liudongpo.tispace
  - host: test1.liudongpo.tispace
  - host: test2.liudongpo.tispace

tikv_servers:
  - host: vsc.liudongpo.tispace
  - host: test1.liudongpo.tispace
  - host: test2.liudongpo.tispace

cdc_servers:
  - host: vsc.liudongpo.tispace
  - host: test1.liudongpo.tispace
  - host: test2.liudongpo.tispace

monitoring_servers:
  - host: vsc.liudongpo.tispace
  - host: test1.liudongpo.tispace
  - host: test2.liudongpo.tispace
  
grafana_servers:
  - host: vsc.liudongpo.tispace
  - host: test1.liudongpo.tispace
  - host: test2.liudongpo.tispace
```
```shell
tiup cluster deploy pr-5302 nightly  topology.yaml -p
tiup cluster patch pr-5302 ./patch/cdc-hotfix-linux-amd64.tar.gz --offline -N test1.liudongpo.tispace:8300,test2.liudongpo.tispace:8300,vsc.liudongpo.tispace:8300
tiup cluster start pr-5302
./cdc cli changefeed create --sink-uri="kafka://172.16.102.16:9092/ticdc-test?protocol=canal-json" --pd http://test1.liudongpo.tispace:2379
./cdc cli changefeed list --pd http://test1.liudongpo.tispace:2379 
 ./cdc cli changefeed pause -c ae859148-8e8b-482d-9887-b6f8cdedc485  --pd http://test1.liudongpo.tispace:2379 
sysbench --db-driver=mysql --mysql-host=test2.liudongpo.tispace --mysql-port=4000 --mysql-user=root --threads=20 --mysql-db=test --tables=100 --table-size=1000000 oltp_write_only prepare
./cdc cli changefeed resume -c ae859148-8e8b-482d-9887-b6f8cdedc485  --pd http://test1.liudongpo.tispace:2379 
```

After:

```log
[2022/05/13 22:11:19.675 +08:00] [INFO] [version.go:47] ["Welcome to Change Data Capture (CDC)"] [release-version=v6.0.0-master-dirty] [git-hash=03a484faba61c7e06c8f443e30f5eb2b32ac3c2f] [git-branch=rustin-patch-buffersink] [utc-build-time="2022-05-13 13:50:39"] [go-version="go version go1.18 darwin/amd64"] [failpoint-build=false]
```

image

image

image

image

image

image

**Comment by hi-rustin** [comment]
Conclusion
Sync elapsed time, before: 34min, after: 24min
kafka sink throughput, before: 47184, after: 49385
                                      before: 7.65 MiB, after: 8.16 MiB
The difference in sink node throughput is not significant.

File: cdc/sink/mq/dispatcher/partition/index_value.go (hunk 2 above)
Comment (hi-rustin) [reply cdc/sink/mq/dispatcher/partition/index_value.go:26]: See: https://github.com/pingcap/tiflow/pull/5302#issuecomment-1126655847

File: cdc/sink/table_sink.go (hunk 3 above)
Comment (hi-rustin) [reply cdc/sink/table_sink.go:150]: For pipeline: https://github.com/pingcap/tiflow/blob/033e7d49c7f21be5722009f12dfe3b5b991b2c1a/cdc/processor/pipeline/sink.go#L288


For actor: https://github.com/pingcap/tiflow/pull/5248/files

**Review by amyangfei — APPROVED** [review]

**Review by CharlesCheung96 — APPROVED** [review]

**Pull request merged**